src/hotspot/share/runtime/thread.cpp

```diff
@@ -1631,38 +1621,35 @@ void JavaThread::check_and_handle_async_exceptions() {
           }
         ls.print_cr(" of type: %s", _pending_async_exception->klass()->external_name());
       }
-      _pending_async_exception = NULL;
-      // Clear condition from _suspend_flags since we have finished processing it.
-      clear_suspend_flag(_has_async_exception);
     }
-  }
+    // Always null out the _pending_async_exception oop here since the async condition was
+    // already cleared above and thus considered handled.
+    _pending_async_exception = NULL;
+  } else {
+    assert(_is_unsafe_access_error, "must be");
+    DEBUG_ONLY(_is_unsafe_access_error = false);
 
-  if (condition == _async_unsafe_access_error && !has_pending_exception()) {
     // We may be at method entry which requires we save the do-not-unlock flag.
     UnlockFlagSaver fs(this);
     switch (thread_state()) {
     case _thread_in_vm: {
       JavaThread* THREAD = this;
       Exceptions::throw_unsafe_access_internal_error(THREAD, __FILE__, __LINE__, "a fault occurred in an unsafe memory access operation");
-      return;
-    }
-    case _thread_in_native: {
-      ThreadInVMfromNative tiv(this);
-      JavaThread* THREAD = this;
-      Exceptions::throw_unsafe_access_internal_error(THREAD, __FILE__, __LINE__, "a fault occurred in an unsafe memory access operation");
-      return;
+      // We might have blocked in a ThreadBlockInVM wrapper in the call above so make sure we process pending
+      // suspend requests and object reallocation operations if any since we might be going to Java after this.
+      SafepointMechanism::process_if_requested_with_exit_check(this, true /* check asyncs */);
```

Comment:
I have to wonder why this isn't handled in the TBIVM?

Reply:
In 8270085 we had to disable by default processing of suspend requests due to possible deadlocks. So although suspend requests are implemented with handshakes, calls to process_if_requested() from ~TBIVM will not process them. As for object reallocation I'm not sure but I think it's just because there is no need to process them until we are about to go back to Java.